Linter: create rule to remove additional items when items is an object

Karan-Palan opened the pull request:
### Implementation

**Rule ID**:  dangling_additional_items
**Description**: "`additionalItems` is ignored when `items` is an object"  
**Condition**: Detects `additionalItems` in JSON Schema objects where `items` is an object (not an array).  
**Transform**: Removes the redundant `additionalItems` property.  
**Dialects**: Supports draft4, draft6, draft7, and 2019-09. Excluded for draft-2020-12 (where `additionalItems` was removed from the spec).  

  - Registered the rule in the **Common rules** section (correctness category).  
  - Added 10 test cases across all relevant JSON Schema drafts.  

### Test Cases

Positive Cases:  Removal of `additionalItems` when `items` is an object.  
Negative Cases:  No change for `items` as an array or when `items` is undefined.  
Edge Cases:  Complex `items` objects (e.g., with `if`/`then` conditions) - correctly handled.  
Context Preservation:  Tests with various other keywords to ensure proper behavior.

jviotti commented on src/extension/alterschema/linter/dangling_additional_items.h:
```diff
@@ -0,0 +1,30 @@
+                "http://json-schema.org/draft-06/schema#",
+                "http://json-schema.org/draft-04/schema#"}) &&
+           schema.is_object() && schema.defines("items") &&
+           schema.defines("additionalItems") && schema.at("items").is_object();
```
Note that this applies is the schema is a booleans too! I also put comments to clarify this in the rule name and description. I think what you want to check for is if the keyword is set to a schema:

```suggestion
           schema.defines("additionalItems") && is_schema(schema.at("items"));
```

jviotti commented on test/alterschema/alterschema_lint_2019_09_test.cc:
```diff
@@ -1661,8 +1661,7 @@ TEST(AlterSchema_lint_2019_09, unnecessary_allof_ref_wrapper_5) {
     "allOf": [
-      { 
-        "type": "integer",
+      { "type": "integer",
```
Unintended change?

jviotti commented:
I left a few comments. On testing:

- Can you test cases where `items` is set to boolean schemas?
- Can you test cases where `items` is set to an array and therefore confirm that `additionalItems` is NOT removed?

jviotti commented on src/extension/alterschema/alterschema.cc:
```diff
@@ -52,6 +52,7 @@ static auto every_item_is_boolean(const T &container) -> bool {
 #include "linter/content_media_type_without_encoding.h"
 #include "linter/content_schema_default.h"
 #include "linter/content_schema_without_media_type.h"
+#include "linter/dangling_additional_items.h"
```
Also, you have to update the `CMakeLists.txt` file to list this file too

Karan-Palan commented on src/extension/alterschema/linter/dangling_additional_items.h:
```diff
@@ -0,0 +1,30 @@
+               {"https://json-schema.org/draft/2019-09/vocab/applicator",
+                "http://json-schema.org/draft-07/schema#",
+                "http://json-schema.org/draft-06/schema#",
+                "http://json-schema.org/draft-04/schema#"}) &&
```
Yes it does, adding tests for it as well

Karan-Palan commented:
- Renamed the rule, class, file, and all references.
- Updated CMakeLists.txt
- Added Draft 3 support with tests.
- Added comprehensive tests for boolean items, array items, object items, and scenarios where the rule should _not_ apply.

@jviotti , the PR is ready for review

jviotti commented on src/extension/alterschema/linter/additional_items_with_schema_items.h:
```diff
@@ -0,0 +1,31 @@
+                "http://json-schema.org/draft-04/schema#",
+                "http://json-schema.org/draft-03/schema#"}) &&
+           schema.is_object() && schema.defines("items") &&
+           schema.defines("additionalItems") && schema.at("items").is_object();
```
```suggestion
           schema.defines("additionalItems") && is_schema(schema.at("items"));
```

jviotti commented on src/extension/alterschema/linter/additional_items_with_schema_items.h:
```diff
@@ -0,0 +1,31 @@
+                "http://json-schema.org/draft-04/schema#",
+                "http://json-schema.org/draft-03/schema#"}) &&
+           schema.is_object() && schema.defines("items") &&
+           schema.defines("additionalItems") && schema.at("items").is_object();
```
Also I think you are missing test cases covering that case: `items` set to a boolean

Karan-Palan commented:
@jviotti , resolved merge conflicts. Any comments?
@relequestual, any other tests that come into your mind?
